test/opt/test_fold_spec_const_op_composite.cpp

```diff
@@ -0,0 +1,922 @@
+    "%v4int_0_1_2_3 = OpConstantComposite %v4int %signed_zero %signed_one %signed_two %signed_three",
+    // clang-format on
+};
+
+// A helper function to transform a piece of assembly code from string format
+// to a vector of instruction strings.
+std::string StripOpNameInstructions(const std::string& str) {
+  std::stringstream ss(str);
+  std::ostringstream oss;
+  std::string inst_str;
+  while (std::getline(ss, inst_str, '\n')) {
+    if (inst_str.find("OpName %") == std::string::npos) {
+      oss << inst_str << '\n';
+    }
+  }
+  return oss.str();
+}
+
+struct FoldSpecConstantOpAndCompositePassTestCase {
+  // Original constants with unfolded spec constants.
+  std::vector<std::string> original;
+  // Expected cosntants after folding.
+  std::vector<std::string> expected;
+};
+
+using FoldSpecConstantOpAndCompositePassTest = PassTest<
+    ::testing::TestWithParam<FoldSpecConstantOpAndCompositePassTestCase>>;
+
+TEST_P(FoldSpecConstantOpAndCompositePassTest, ParamTestCase) {
+  AssemblyBuilder test_code_builder, expected_code_builder;
+  auto& tc = GetParam();
+  test_code_builder.AppendTypesConstantsGlobals(CommonTypesAndConstants);
+  test_code_builder.AppendTypesConstantsGlobals(tc.original);
+  expected_code_builder.AppendTypesConstantsGlobals(CommonTypesAndConstants);
+  expected_code_builder.AppendTypesConstantsGlobals(tc.expected);
+  std::string original = test_code_builder.GetCode();
+  std::string expected = expected_code_builder.GetCode();
+
+  // Run the optimization and get the result code in disassembly.
+  std::string optimized;
+  bool modified = false;
+  std::tie(optimized, modified) =
+      SinglePassRunAndDisassemble<opt::FoldSpecConstantOpAndCompositePass>(
+          original, /* skip_nop = */ true);
+
+  // Check the optimized code, but ignore the OpName instructions.
+  EXPECT_EQ(
+      StripOpNameInstructions(expected) != StripOpNameInstructions(original),
+      modified);
+  EXPECT_EQ(StripOpNameInstructions(expected),
+            StripOpNameInstructions(optimized));
+}
+
+// Tests that OpSpecConstantComposite opcodes are replace with
+// OpConstantComposite correctly.
+INSTANTIATE_TEST_CASE_P(
+    CompositeVector, FoldSpecConstantOpAndCompositePassTest,
+    ::testing::ValuesIn(std::vector<
+                        FoldSpecConstantOpAndCompositePassTestCase>({
+        // clang-format off
+            {
+              // original
+              {
+                "%spec_v2bool = OpSpecConstantComposite %v2bool %bool_true %bool_false",
+                "%spec_v2uint = OpSpecConstantComposite %v2uint %unsigned_one %unsigned_one",
+                "%spec_v2int_a = OpSpecConstantComposite %v2int %signed_one %signed_two",
+                // Spec constants whose value can not be fully resolved should
+                // not be processed.
+                "%spec_int = OpSpecConstant %int 99",
+                // "%spec_v2int_b = OpSpecConstantComposite %v2int %signed_one %spec_int",
```

Comment:
Is this commented line to be preserved somehow? Otherwise, should just delete it.